File: test_modules/guest_modules/go/callbacks.go
package guest

import "fmt"

type StringTransformer func(string) string

func CallTransformer(transformer StringTransformer, value string) string {
	return transformer(value)
}

func ReturnTransformer(suffix string) StringTransformer {
	return func(value string) string {
		return value + suffix
	}
}

func CallFunction(function func(string) int, value string) int {
	return function(value)
}

func CallTransformerWithError(cb func(string) (string, error), value string) (string, error) {
	return cb(value)
}

func CallTransformerRecover(cb func(string) string, value string) (res string, err error) {
	defer func() {
		if r := recover(); r != nil {
			err = fmt.Errorf("callback panic: %v", r)
		}
	}()
	return cb(value), nil
}

File: test_modules/guest_modules/java/src/guest/Callbacks.java
package guest;

import java.util.function.Function;

public class Callbacks {
	public interface StringTransformer {
		String transform(String value);
	}

	public static String callTransformer(StringTransformer transformer, String value) {
		return transformer.transform(value);
	}

	public static StringTransformer returnTransformer(String suffix) {
		return value -> value + suffix;
	}

	public static Integer callFunction(Function<String, Integer> function, String value) {
		return function.apply(value);
	}
}
